Always consider highest stored block as trusted block.

afck opened the pull request:
Even if a trusted hash is specified in the config file, fast sync will now use the highest block from storage as a starting point if that is more recent.

So e.g. a node that started originally with block 1000 as a trusted block, and was then restarted after block 2000, with the same config file, will now start syncing from block 2000 forward, not 1000, without loading the intermediate 1000 headers from storage first.

Closes #3288.

afck committed:
Always consider highest stored block as trusted block.

Even if a trusted hash is specified in the config file, fast sync will
now use the highest block from storage as a starting point if that is
more recent.

So e.g. a node that started originally with block 1000 as a trusted
block, and was then restarted after block 2000, with the same config
file, will now start syncing from block 2000 forward, not 1000, without
loading the intermediate 1000 headers from storage first.

afck commented:
bors merge

fizyk20 commented on node/src/components/chain_synchronizer/operations.rs:
```diff
@@ -146,17 +146,29 @@ where
-                    debug!("no highest block header found in storage");
-                    return Ok(None);
+            (Some(config_header), Some(highest_header)) => {
+                if config_header.height() > highest_header.height() {
```
I'd make it `>=`, so that the trusted hash takes priority if both are at the same height - I think that would be more correct. Not that important, though.

afck replied:
Hmm… if they're at the same height but different, I think the right thing to do would be to return an error. Maybe we can do that in another PR?

fizyk20 replied:
That's also an option, and probably the best one right now, actually. But if we ever decide to support correcting being on a wrong fork of the chain, it might be easier to just prioritize the trusted hash.

I'm fine with doing it in another PR :+1: I could even do it in #3280 (I'm going to merge yours into it once it's in `dev`).

afck committed:
Add info log if using stored block instead of configured trusted hash.

afck committed:
Error if trusted hash is from different fork.

The chain synchronizer will now return an error if the trusted block
from the config file is different from a stored block at the same
height.

afck commented:
bors try

afck commented:
bors merge
